src/arithmetic/arithmetic_expression.c

```diff
@@ -233,32 +247,34 @@ SIValue AR_EXP_Evaluate(AR_ExpNode *root, const Record r) {
 			}
 
 			/* Validate before evaluation. */
-			char *error;
-			if(!_AR_EXP_ValidateInvocation(root->op.f, sub_trees, root->op.child_count, &error)) {
-				// The expression tree failed its validations and prepared an error message.
+			if(!_AR_EXP_ValidateInvocation(root->op.f, sub_trees, root->op.child_count)) {
+				// The expression tree failed its validations and set an error message.
 				// Free all associated memory.
 				for(int child_idx = 0; child_idx < root->op.child_count; child_idx++) {
-					SIValue_Free(sub_trees + child_idx);
+					SIValue_Free(&sub_trees[child_idx]);
 				}
-				QueryCtx_SetError(error); // Set the query-level error.
-
-				// Invoke the exception handler, exiting this routine and returning to
-				// the point on the stack where the handler was instantiated.
-				jmp_buf *env = QueryCtx_GetExceptionHandler();
-				longjmp(*env, 1);
+				_RaiseException(); // Raise an exception.
 			}
 			/* Evaluate self. */
 			result = root->op.f->func(sub_trees, root->op.child_count);
-			/* Free any SIValues that were allocated while evaluating this tree. */
+			/* Free any SIValues that were allocated while evaluating this tree.
+			 * for example 'a'+'b'+'c'+'d' will evalute to
+			          +(abcd)
+			         /       \
+			     +(ab)      +(cd)
+			    /   \       /   \
+			   a     b     c     d
+			   the expressions 'ab', 'cd' which are calculated temporaty values
+			   will be released once the calculation of 'abcd' is done
+			*/
 			for(int child_idx = 0; child_idx < root->op.child_count; child_idx++) {
 				SIValue_Free(&sub_trees[child_idx]);
 			}
 			if(SIValue_IsNull(result) && QueryCtx_EncounteredError()) {
```

Comment:
isn't that enough to check for `QueryCtx_EncounteredError()` why do we care about `result` type here?

Reply:
We don't; this is theoretically an optimization on the assumption that `SIValue_IsNull` is a cheaper call than `QueryCtx_EncounteredError`, and that null values are relatively unlikely in the wild. (If we've set an error, we always return null.)